From: Stanislaw Kardach <stanislaw.kardach@caviumnetworks.com>
Date: Tue, 22 Mar 2016 12:05:09 +0100
Subject: [PATCH] nova: Update vga, console defaults for armv7, aarch64

Nova hardcodes default options for both video=cirrus and console.
armv7 and aarch64 VMs require video=vga, and since most applications
use PL011 serial driver for guests, adding console=ttyAMA0 is also
a nice UX addition.
---
 .../openstack/files/nova-libvirt-vga-console.patch | 39 ++++++++++++++++++++++
 deployment/puppet/openstack_tasks/manifests/roles/compute.pp   | 19 +++++++++++
 2 files changed, 58 insertions(+)
 create mode 100644 deployment/puppet/openstack/files/nova-libvirt-vga-console.patch

diff --git a/deployment/puppet/openstack/files/nova-libvirt-vga-console.patch b/deployment/puppet/openstack/files/nova-libvirt-vga-console.patch
new file mode 100644
index 0000000..c5f8e5a
--- /dev/null
+++ b/deployment/puppet/openstack/files/nova-libvirt-vga-console.patch
@@ -0,0 +1,40 @@
+From: Stanislaw Kardach <stanislaw.kardach@caviumnetworks.com>
+Date: Tue, 22 Mar 2016 12:05:09 +0100
+Subject: [PATCH] Update vga, console defaults for armv7 and aarch64
+
+Nova hardcodes default options for both video=cirrus and console.
+armv7 and aarch64 VMs require video=vga, and since most applications
+use PL011 serial driver for guests, adding console=ttyAMA0 is also
+a nice UX addition.
+
+Signed-off-by: Stanislaw Kardach <stanislaw.kardach@caviumnetworks.com>
+Signed-off-by: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
+---
+
+diff --git a/virt/libvirt/driver.py b/virt/libvirt/driver.py
+index 920a283..751b140 100644
+--- a/virt/libvirt/driver.py
++++ b/virt/libvirt/driver.py
+@@ -3834,7 +3834,11 @@
+         if virt_type == "xen":
+             guest.os_cmdline = "ro root=%s" % root_device_name
+         else:
++            guestarch = libvirt_utils.get_arch(image_meta)
+             guest.os_cmdline = ("root=%s %s" % (root_device_name, CONSOLE))
++            if guestarch in (arch.ARMV7, arch.AARCH64):
++                # NOTE(armband): ARM v7/v8 use PL011 drv, add ttyAMA0 console
++                guest.os_cmdline += " console=ttyAMA0"
+             if virt_type == "qemu":
+                 guest.os_cmdline += " no_timer_check"
+         if instance.ramdisk_id:
+@@ -3972,7 +3972,9 @@ class LibvirtDriver(driver.ComputeDriver):
+             video.type = 'xen'
+         elif CONF.libvirt.virt_type == 'parallels':
+             video.type = 'vga'
+-        elif guestarch in (arch.PPC, arch.PPC64, arch.PPC64LE):
++        elif guestarch in (arch.ARMV7, arch.AARCH64,
++                           arch.PPC, arch.PPC64, arch.PPC64LE):
++            # NOTE(armband): Added ARM v7/v8, same as on PPC.
+             # NOTE(ldbragst): PowerKVM doesn't support 'cirrus' be default
+             # so use 'vga' instead when running on Power hardware.
+             video.type = 'vga'
diff --git a/deployment/puppet/openstack_tasks/manifests/roles/compute.pp b/deployment/puppet/openstack_tasks/manifests/roles/compute.pp
index c0349dc..46b1801 100644
--- a/deployment/puppet/openstack_tasks/manifests/roles/compute.pp
+++ b/deployment/puppet/openstack_tasks/manifests/roles/compute.pp
@@ -157,4 +157,10 @@ class openstack::compute (

   include ::nova::params
+
+  if ! defined(Package['patch']) {
+    package { 'patch':
+      ensure => 'present',
+    }
+  }
 
   case $::osfamily {
@@ -250,3 +256,4 @@ class openstack::compute (
 
   $notify_on_state_change = 'vm_and_task_state'
+  $nova_path = '/usr/lib/python2.7/dist-packages/nova'
 
@@ -273,5 +280,17 @@ class openstack::compute (
     memcached_servers                  => $memcached_addresses,
     cinder_catalog_info                => pick($nova_hash_real['cinder_catalog_info'], 'volumev2:cinderv2:internalURL'),
     rabbit_heartbeat_timeout_threshold => $::os_service_default,
+  } ->
+  # FIXME(armband): Workaround for missing arm defaults in nova libvirt driver
+  file { "${nova_path}/libvirt-vga-console.patch":
+    ensure => "file",
+    source => "puppet:///modules/openstack/nova-libvirt-vga-console.patch",
+  } ->
+  exec { 'nova libvirt driver patch arm defaults':
+    path    => ['/usr/bin'],
+    command => "patch -p1 < ${nova_path}/libvirt-vga-console.patch",
+    unless  => "patch -p1 -R -N --dry-run < ${nova_path}/libvirt-vga-console.patch",
+    cwd     => $nova_path,
+    require => [Package['patch']],
   }
 
